### PR TITLE
Blackplum a.k.a IMKG68 - new 68% mechanical keyboard

### PR DESCRIPTION
**Pull request by eriqadams**
## Description
Blackplum a.k.a IMKG68 - new 68% mechanical keyboard

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [X] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/blackplum/blackplum.h [hunk 1]
```diff
@@ -0,0 +1,22 @@
+#pragma once
+
+#include "quantum.h"
+
+#define LAYOUT_68( \
```
Comment (fauxpark) [inline keyboards/blackplum/blackplum.h]:
```suggestion
#define LAYOUT_68_ansi( \
```

File: keyboards/blackplum/config.h [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+#define MATRIX_ROWS 9
+#define MATRIX_COLS 8
+
+/* key matrix pins */
+#define MATRIX_ROW_PINS { C6, B6, B4, B5, D6, D7, D5, D3, D4 }
+#define MATRIX_COL_PINS { D0, D1, D2, F7, F6, F5, F4, F1 }
+#define UNUSED_PINS
+
+/* Set 0 if debouncing isn't needed */
+#define DEBOUNCING_DELAY 5
```
Comment (fauxpark) [inline keyboards/blackplum/config.h]:
```suggestion
#define DEBOUNCE 5
```

File: keyboards/blackplum/config.h [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+/* Mechanical locking support. Use KC_LCAP, KC_LNUM or KC_LSCR instead in keymap */
+#define LOCKING_SUPPORT_ENABLE
+
+/* Locking resynchronize hack */
+#define LOCKING_RESYNC_ENABLE
+
+/* key combination for command */
+#define IS_COMMAND() ( \
+    keyboard_report->mods == (MOD_BIT(KC_LSHIFT) | MOD_BIT(KC_RSHIFT)) \
+)
```
Comment (fauxpark) [inline keyboards/blackplum/config.h]: This block should be removed, as it's the default in QMK.

File: keyboards/blackplum/info.json [hunk 4]
```diff
@@ -0,0 +1,12 @@
+{
+    "keyboard_name": "foobar",
```
Comment (fauxpark) [inline keyboards/blackplum/info.json]:
```suggestion
    "keyboard_name": "Blackplum",
```

File: keyboards/blackplum/info.json [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+    "url": "",
+    "maintainer": "qmk",
+    "width": 15,
+    "height": 5,
+    "layouts": {
+      "LAYOUT_60_ansi": {
```
Comment (fauxpark) [inline keyboards/blackplum/info.json]:
```suggestion
      "LAYOUT_68_ansi": {
```

File: keyboards/blackplum/keymaps/default/keymap.c [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+#define _FN 2
+
+
+
+enum custom_keycodes {
+  QWERTY = SAFE_RANGE,
+};
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+  [_BASE] = LAYOUT_68(
```
Comment (fauxpark) [inline keyboards/blackplum/keymaps/default/keymap.c]:
```suggestion
  [_BASE] = LAYOUT_68_ansi(
```

File: keyboards/blackplum/keymaps/default/keymap.c [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    KC_GESC, 	KC_1,    	KC_2,    	KC_3,    KC_4,    	KC_5,    KC_6,    KC_7,    	KC_8,    	KC_9,    	KC_0,    	KC_MINS, 	KC_EQL,  	KC_BSPC, 	KC_INS,		KC_PGUP,\
+    KC_TAB,  	KC_Q,    	KC_W,    	KC_E,    KC_R,    	KC_T,    KC_Y,    KC_U,    	KC_I,    	KC_O,    	KC_P,    	KC_LBRC, 	KC_RBRC, 	KC_BSLS, 	KC_DEL,		KC_PGDOWN,\
+    KC_CAPS, 	KC_A,    	KC_S,    	KC_D,    KC_F,    	KC_G,    KC_H,    KC_J,    	KC_K,    	KC_L,    	KC_SCLN, 	KC_QUOT, 				KC_ENT,\
+    KC_LSFT, 	KC_Z,    	KC_X,    	KC_C,    KC_V,    	KC_B,    KC_N,    KC_M,    	KC_COMM, 				KC_DOT,  	KC_SLSH, 				KC_RSFT,	KC_UP,\
+    KC_LCTL , 	KC_LGUI , 	KC_LALT ,                      	KC_SPC ,											KC_RALT , 	MO(_FN) ,	KC_RCTL, 	KC_LEFT,	KC_DOWN,	KC_RIGHT
+  ),
+  
+  [_WINLCK] = LAYOUT_68(
```
Comment (fauxpark) [inline keyboards/blackplum/keymaps/default/keymap.c]:
```suggestion
  [_WINLCK] = LAYOUT_68_ansi(
```

File: keyboards/blackplum/keymaps/default/keymap.c [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    KC_GESC, 	KC_1,    	KC_2,    	KC_3,    KC_4,    	KC_5,    KC_6,    KC_7,    	KC_8,    	KC_9,    	KC_0,    	KC_MINS, 	KC_EQL,  	KC_BSPC, 	KC_INS,		KC_PGUP,\
+    KC_TAB,  	KC_Q,    	KC_W,    	KC_E,    KC_R,    	KC_T,    KC_Y,    KC_U,    	KC_I,    	KC_O,    	KC_P,    	KC_LBRC, 	KC_RBRC, 	KC_BSLS, 	KC_DEL,		KC_PGDOWN,\
+    KC_CAPS, 	KC_A,    	KC_S,    	KC_D,    KC_F,    	KC_G,    KC_H,    KC_J,    	KC_K,    	KC_L,    	KC_SCLN, 	KC_QUOT, 				KC_ENT,\
+    KC_LSFT, 	KC_Z,    	KC_X,    	KC_C,    KC_V,    	KC_B,    KC_N,    KC_M,    	KC_COMM, 				KC_DOT,  	KC_SLSH, 				KC_RSFT,	KC_UP,\
+    KC_LCTL , 	KC_NO , 	KC_LALT ,                      	KC_SPC ,											KC_RALT , 	MO(_FN) ,	KC_RCTL, 	KC_LEFT,	KC_DOWN,	KC_RIGHT
+  ),
+
+  [_FN] = LAYOUT_68(
```
Comment (fauxpark) [inline keyboards/blackplum/keymaps/default/keymap.c]:
```suggestion
  [_FN] = LAYOUT_68_ansi(
```

File: keyboards/blackplum/readme.md [hunk 9]
```diff
@@ -0,0 +1,4 @@
+
```
Comment (fauxpark) [inline keyboards/blackplum/readme.md]:
```suggestion
```

File: keyboards/blackplum/readme.md [hunk 10]
```diff
@@ -0,0 +1,4 @@
...
+## Blackplum 68% keyboard with RGB underglow.
```
Comment (fauxpark) [inline keyboards/blackplum/readme.md]:
```suggestion
# Blackplum 68% keyboard with RGB underglow.
```

File: keyboards/blackplum/rules.mk [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+#     at the end, this will be done automatically to create a 32-bit value in your
+#     source code.
+#
+#     If no clock division is performed on the input clock inside the AVR (via the
+#     CPU clock adjust registers or the clock division fuses), this will be equal to F_CPU.
+F_USB = $(F_CPU)
+
+# Interrupt driven control endpoint task(+60)
+OPT_DEFS += -DINTERRUPT_CONTROL_ENDPOINT
+
```
Comment (fauxpark) [inline keyboards/blackplum/rules.mk]: Please remove these lines; QMK provides them as default.

File: keyboards/blackplum/rules.mk [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+
+# Bootloader selection
+#   Teensy       halfkay
+#   Pro Micro    caterina
+#   Atmel DFU    atmel-dfu
+#   LUFA DFU     lufa-dfu
+#   QMK DFU      qmk-dfu
+#   atmega32a    bootloadHID
```
Comment (fauxpark) [inline keyboards/blackplum/rules.mk]:
```suggestion
#   ATmega32A    bootloadHID
#   ATmega328P   USBasp
```

File: keyboards/blackplum/rules.mk [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+# If you don't know the bootloader type, then you can specify the
+# Boot Section Size in *bytes* by uncommenting out the OPT_DEFS line
+#   Teensy halfKay      512
+#   Teensy++ halfKay    1024
+#   Atmel DFU loader    4096
+#   LUFA bootloader     4096
+#   USBaspLoader        2048
+# OPT_DEFS += -DBOOTLOADER_SIZE=4096
+
+
```
Comment (fauxpark) [inline keyboards/blackplum/rules.mk]: As above.

File: keyboards/blackplum/rules.mk [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+EXTRAKEY_ENABLE = yes	  # Audio control and System control(+450)
+CONSOLE_ENABLE = no			# Console for debug(+400)
+COMMAND_ENABLE = no     # Commands for debug and configuration
+SLEEP_LED_ENABLE = no   # Breathing sleep LED during USB suspend
+NKRO_ENABLE = yes				# USB Nkey Rollover - if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+BACKLIGHT_ENABLE = no  # Enable keyboard backlight functionality
+AUDIO_ENABLE = no
+RGBLIGHT_ENABLE = yes
+
+LAYOUTS = 68
```
Comment (fauxpark) [inline keyboards/blackplum/rules.mk]:
```suggestion
LAYOUTS = 68_ansi
```

File: keyboards/blackplum/rules.mk [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+#   Atmel DFU loader    4096
+#   LUFA bootloader     4096
+#   USBaspLoader        2048
+# OPT_DEFS += -DBOOTLOADER_SIZE=4096
+
+
+# Build Options
+#   comment out to disable the options.
+#
+BOOTMAGIC_ENABLE = yes	# Virtual DIP switch configuration(+1000)
```
Comment (fauxpark) [inline keyboards/blackplum/rules.mk]:
```suggestion
BOOTMAGIC_ENABLE = yes	# Virtual DIP switch configuration
```

File: keyboards/blackplum/rules.mk [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+MOUSEKEY_ENABLE = yes	  # Mouse keys(+4700)
```
Comment (fauxpark) [inline keyboards/blackplum/rules.mk]:
```suggestion
MOUSEKEY_ENABLE = yes	  # Mouse keys
```

File: keyboards/blackplum/rules.mk [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+EXTRAKEY_ENABLE = yes	  # Audio control and System control(+450)
```
Comment (fauxpark) [inline keyboards/blackplum/rules.mk]:
```suggestion
EXTRAKEY_ENABLE = yes	  # Audio control and System control
```

File: keyboards/blackplum/rules.mk [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+CONSOLE_ENABLE = no			# Console for debug(+400)
```
Comment (fauxpark) [inline keyboards/blackplum/rules.mk]:
```suggestion
CONSOLE_ENABLE = no			# Console for debug
```

File: keyboards/blackplum/readme.md [hunk 19]
```diff
@@ -0,0 +1,4 @@
...
+
+![Blackplum layout](https://i.imgur.com/EOUQk4J.png)
```
Comment (fauxpark) [inline keyboards/blackplum/readme.md]: Please also format this file like the template:

https://github.com/qmk/qmk_firmware/blob/master/quantum/template/avr/readme.md

File: keyboards/blackplum/keymaps/default/keymap.c [hunk 20]
```diff
@@ -0,0 +1,43 @@
...
+    KC_GRV, 	KC_F1,   	KC_F2,   	KC_F3,   KC_F4,   	KC_F5,   KC_F6,   KC_F7,   	KC_F8,   	KC_F9,   	KC_F10,  	KC_F11,  	KC_F12,  	_______,	KC_MPLY,	KC_HOME,\
+    _______, 	RGB_MOD, 	RGB_HUI,   	RGB_SAI, RGB_VAI, 	_______, _______, _______, 	_______, 	_______, 	KC_PSCR, 	KC_SLCK, 	KC_PAUSE,  	_______,	KC_MUTE,	KC_END,\
+    _______,	RGB_TOG, 	RGB_HUD, 	RGB_SAD, RGB_VAD, 	_______, _______, _______, 	_______, 	_______, 	_______, 	_______, 				_______,\
+    _______,	_______, 	_______, 	_______, _______, 	_______, _______, _______,	_______, 				_______, 	_______, 				_______,	KC_VOLU,\
+    _______, 	TG(_WINLCK),_______, 						RESET, 												_______,	_______,	_______,	KC_MPRV,    KC_VOLD, 	KC_MNXT
```
Comment (fauxpark) [inline keyboards/blackplum/keymaps/default/keymap.c]: You could do away with the `_WINLCK` layer by replacing the `TG()` with a custom keycode to toggle the GUI keys:
```c
enum custom_keycodes {
    TG_GUI = SAFE_RANGE
};

bool process_record_user(uint16_t keycode, keyrecord_t *record) {
    if(record->event.pressed) {
        switch(keycode) {
            case TG_GUI:
                keymap_config.no_gui ^= 1;
                return false;
        }
    }
    return true;
}
```

File: keyboards/blackplum/info.json [hunk 21]
```diff
@@ -0,0 +1,12 @@
...
...5}, {"label":"Caps Lock", "x":0, "y":2, "w":1.75}, {"label":"A", "x":1.75, "y":2}, {"label":"S", "x":2.75, "y":2}, {"label":"D", "x":3.75, "y":2}, {"label":"F", "x":4.75, "y":2}, {"label":"G", "x":5.75, "y":2}, {"label":"H", "x":6.75, "y":2}, {"label":"J", "x":7.75, "y":2}, {"label":"K", "x":8.75, "y":2}, {"label":"L", "x":9.75, "y":2}, {"label":":", "x":10.75, "y":2}, {"label":"\"", "x":11.75, "y":2}, {"label":"Enter", "x":12.75, "y":2, "w":2.25}, {"label":"Shift", "x":0, "y":3, "w":2.25}, {"label":"Z", "x":2.25, "y":3}, {"label":"X", "x":3.25, "y":3}, {"label":"C", "x":4.25, "y":3}, {"label":"V", "x":5.25, "y":3}, {"label":"B", "x":6.25, "y":3}, {"label":"N", "x":7.25, "y":3}, {"label":"M", "x":8.25, "y":3}, {"label":"<", "x":9.25, "y":3}, {"label":">", "x":10.25, "y":3}, {"label":"?", "x":11.25, "y":3}, {"label":"Shift", "x":12.25, "y":3, "w":2.75}, {"label":"Ctrl", "x":0, "y":4, "w":1.25}, {"label":"Win", "x":1.25, "y":4, "w":1.25}, {"label":"Alt", "x":2.5, "y":4, "w":1.25}, {"x":3.75, "y":4, "w":6.25}, {"label":"Alt", "x":10, "y":4, "w":1.25}, {"label":"Win", "x":11.25, "y":4, "w":1.25}, {"label":"Menu", "x":12.5, "y":4, "w":1.25}, {"label":"Ctrl", "x":13.75, "y":4, "w":1.25}]
```
Comment (fauxpark) [inline keyboards/blackplum/info.json]: This layout is missing some keys and is actually a 60%, please update it.

File: keyboards/blackplum/keymaps/default/keymap.c (hunk 20 above)
Comment (eriqadams) [reply keyboards/blackplum/keymaps/default/keymap.c]: okay i will update it

File: keyboards/blackplum/info.json (hunk 21 above)
Comment (eriqadams) [reply keyboards/blackplum/info.json]: thanks i forgot to change

File: keyboards/blackplum/keymaps/default/keymap.c (hunk 20 above)
Comment (eriqadams) [reply keyboards/blackplum/keymaps/default/keymap.c]: done. updated

File: keyboards/blackplum/blackplum.h (hunk 1 above)
Comment (eriqadams) [reply keyboards/blackplum/blackplum.h]: done updated.

File: keyboards/blackplum/config.h (hunk 2 above)
Comment (eriqadams) [reply keyboards/blackplum/config.h]: Done Updated.

File: keyboards/blackplum/info.json (hunk 4 above)
Comment (eriqadams) [reply keyboards/blackplum/info.json]: Done.

File: keyboards/blackplum/info.json (hunk 5 above)
Comment (eriqadams) [reply keyboards/blackplum/info.json]: Done.

File: keyboards/blackplum/rules.mk (hunk 18 above)
Comment (eriqadams) [reply keyboards/blackplum/rules.mk]: done

File: keyboards/blackplum/readme.md (hunk 19 above)
Comment (eriqadams) [reply keyboards/blackplum/readme.md]: Done.

File: keyboards/blackplum/info.json (hunk 21 above)
Comment (eriqadams) [reply keyboards/blackplum/info.json]: Done.

File: keyboards/blackplum/config.h [hunk 22]
```diff
@@ -0,0 +1,39 @@
+#pragma once
+
+#include "config_common.h"
+
+/* USB Device descriptor parameter */
+#define VENDOR_ID       0x0099
+#define PRODUCT_ID      0x6800
```
Comment (drashna) [inline keyboards/blackplum/config.h]: Is this an unused ID?

File: keyboards/blackplum/keymaps/default/keymap.c [hunk 23]
```diff
@@ -0,0 +1,45 @@
+#include QMK_KEYBOARD_H
+
+extern keymap_config_t keymap_config;
```
Comment (drashna) [inline keyboards/blackplum/keymaps/default/keymap.c]: This line is not actually needed, and should be removed. 

```suggestion
```

Comment (eriqadams) [reply keyboards/blackplum/keymaps/default/keymap.c]: Done

File: keyboards/blackplum/config.h (hunk 22 above)
Comment (eriqadams) [reply keyboards/blackplum/config.h]: VENDOR_ID  0x0099 according to https://usb-ids.gowdy.us/read/UD?restrict=0 is unused ID, should I leave to default 0xFEED ?

**Comment by noroadsleft** [comment]
@eriqadams For future reference, we recommend against committing to the master branch, because of the potential for merge conflicts with later changes. For QMK development of any sort, we highly recommend keeping your master updated, but **NEVER** committing to it. Instead, changes should be done in branches (which are basically free in Git) and PRs issued from those branches when you're developing.

There are instructions on how to keep your master updated here:

[**Best Practices: Your fork's master: Update Often, Commit Never**](https://docs.qmk.fm/#/newbs_best_practices?id=your-fork39s-master-update-often-commit-never)

If you need any help with this just ask.

This PR is fine, but you'll need to synchronize your `master` branch after I merge your PR... right now!

Thanks!

**Pull request merged**